index.js

```diff
@@ -116,33 +92,146 @@ function getConstructorName(errorLike) {
 }
 
 /**
- * ### .getMessage(errorLike)
+ * ### .describeExpectedError(criteria)
  *
- * Gets the error message from an error.
- * If `err` is a String itself, we return it.
- * If the error has no message, we return an empty string.
+ * Return a string describing what kind of `Error` instance is expected based on
+ * `criteria` (which is an object returned by `.createCriteria`).
  *
- * @name getMessage
- * @param {Error|String} errorLike
- * @namespace Utils
- * @api public
+ * @param {Object} criteria
+ * @returns String
  */
+function describeExpectedError(criteria) {
+  if (criteria.errLike instanceof Error) {
+    // Mimic `util.inspect`.
+    return '[' + criteria.errLike.toString() + ']';
+  }
 
-function getMessage(errorLike) {
-  var msg = '';
-  if (errorLike && errorLike.message) {
-    msg = errorLike.message;
-  } else if (typeof errorLike === 'string') {
-    msg = errorLike;
+  var desc = '';
+  var constructor = getConstructorName(criteria.errLike);
+  var first = constructor.charAt(0).toLowerCase();
+  var vowels = [ 'a', 'e', 'i', 'o', 'u' ];
+  var article = vowels.indexOf(first) > -1 ? 'an ' : 'a '; // eslint-disable-line no-magic-numbers
+  desc += article + constructor;
+
+  if (criteria.errMsgMatcherType === 'string') {
+    desc += ' including \'' + criteria.errMsgMatcher + '\'';
+  } else if (criteria.errMsgMatcherType === 'regexp') {
+    desc += ' matching ' + criteria.errMsgMatcher;
   }
 
-  return msg;
+  return desc;
 }
 
+/**
+ * ### .checkError(errObj, criteria)
+ *
+ * Validate that `errObj` is an `Error` instance that matches `criteria` (which
+ * is an object returned by `.createCriteria`).
+ *
+ * @param {Error} errObj
+ * @param {Object} criteria
+ * @returns Boolean
+ */
+function checkError(errObj, criteria) {
+  if (criteria.isErrLikeInstanceOfError && errObj !== criteria.errLike) {
+    return false;
+  }
+
+  if (criteria.isErrLikeErrorConstructor &&
+      !(errObj instanceof criteria.errLike)) {
+    return false;
+  }
+
+  if ((criteria.errMsgMatcherType === 'string' ||
+      criteria.errMsgMatcherType === 'regexp') &&
+      !compatibleMessage(errObj, criteria.errMsgMatcher)) {
```

Comment:
As I've stated before, what happens if `errObj` is a string?

IMO if we plan on moving most of the logic to this module instead of keeping it in the core I think it would be suitable for us to deal with throwing Strings here too.

Reply:
This seems like a good reason to not support those.

Let's just not forget about updating our docs on the core then.